memcached: Add NixOS support

ocharles opened the pull request.

ocharles committed:
memcached: Add NixOS support

offlinehacker commented:
Isn't here the preferred method(for ex. like postgres does) to create extra user and group by default, or is there any other reason why memchached user should be put in global nixos config?

ocharles commented:
@offlinehacker not sure I understand your question - I do set a default user ('memcached`), but give users the option to use a different user.

offlinehacker commented:
Sorry i missed that part, my mistake. Then it should be just fine :)

Phreedom commented:
This again probably at least need support for basic config. The default memcached config seems to bind to 0.0.0.0 and is a very bad idea :(

ocharles commented:
Could you give me pointers on things that people would normally want to configure?

offlinehacker commented:
Something like listen ip, port, user, maxconn, cachesize, and additional options(you could also just leave maxconn and cachesize to additional options). For description http://techgurulive.com/2010/01/26/how-to-configure-memcached-memcached-configuration-parameters/

Phreedom commented:
Basically, a good idea is to let people directly append stuff to the config file, or if there's too much stuff being passed via cli, to some "extra cli options" thing.

Also the service shouldn't expose the user to attacks, leaks etc if simply enabled without any options set. So for apache, it might be a good idea to bind to all interfaces by default, since it doesn't let anyone modify anything and doesn't publish anything useful by default. memcached on the other hand rarely seems to be used as a public service, so letting the world read and white your cache without any kind of auth doesn't sound like a good idea.

domenkozar commented:
I'd love to see this one merged once we have basic config support :)

ocharles commented:
I'll see if I can get this finished later today.

ocharles committed:
memcached: Add more options and extraOptions support

ocharles commented:
Ok! Rebased the branch on top of master, and added another commit that adds some basic configuration options, and provides `extraOptions` support. Hopefully this is enough to get into master, but if you have more suggestions I'm happy to incorporate them.

domenkozar commented on modules/services/databases/memcached.nix:
```diff
@@ -0,0 +1,97 @@
+    users.extraUsers = singleton
+      { name = cfg.user;
+        description = "Memcached server user";
+      };
```
You probably want to assert "user == memcached" before adding that user. Normally user already exists if someone want's it to be something else than memcached

offlinehacker replied:
Can you give an example what do you mean? I coudn't find in any existing nixos expression and @ocharles seems confused too.

domenkozar replied:
https://github.com/NixOS/nixos/blob/master/modules/services/web-servers/nginx/default.nix#L77

offlinehacker replied:
I don't think that's the only way. Imagine if i want to have multiple isolated memchached instances running(i can hardly imagine when would you want something like that, but let's say that someone wants). It becomes confusing for someone when users are created only for first instance. I think it's better to have an option to specify if you want to create a user, or just leave everything alone. You can always set users.extraUsers.memchached.createUser to false and user won't be created on a system and get rid of additional unecessary complexity. That's probably also the case why this is not checked elsewhere. While the idea is good, i think it should be handled on another way.

domenkozar commented:
One minor thing, otherwise I think it's ready for merge!

Phreedom commented:
Looks good. Please look at how other services which let you pick the user handle it. I'm too lazy to do it mylsef :)

ocharles commented:
I'm afraid I don't understand what needs to be done with the user stuff. I looked at `postgresql` which hard codes it, and https://github.com/NixOS/nixos/blob/master/modules/services/databases/mongodb.nix does the same as what I'm doing, as does MySQL. What are you guys looking for?

peti commented:
IMHO, more sophisticated username-configuration mechanisms can be added later, when the need actually arises. Personally, I am quite happy with the module as it is now. Thank you for your efforts, Oliver!
